### DISCUSSION TITLE
How can I compile an AdaptiveCpp app which doesn't output AdaptiveCpp errors to console?

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1559] [Q&A]
Hello,

How can I compile an AdaptiveCpp app which doesn't output AdaptiveCpp errors to console? I know I can set ACPP_DEBUG_LEVEL=0 environment variable, but I need this not for personal use but for a commercial app.

Previously, when I didn't want AdaptiveCpp to output warnings, I compiled AdaptiveCpp with ` -DHIPSYCL_DEBUG_LEVEL=1`, and it worked. However, I just updated AdaptiveCpp and tried to recompile it with ` -DHIPSYCL_DEBUG_LEVEL=0`, but error messages are still generated, particularly, `[AdaptiveCpp Error] from D:/source/AdaptiveCpp/src/runtime/cuda/cuda_allocator.cpp:31 @ allocate(): cuda_allocator: cudaMalloc() failed (error code = CUDA:2)` message.

I am handling and logging all errors manually, so I don't need them to appear in console.

**Comment by illuhad** [comment]
Can you please try `-DACPP_DEBUG_LEVEL=...` instead of `-DHIPSYCL_DEBUG_LEVEL`? That option was recently renamed, and I'm not sure whether we put backwards compatibility logic in place for this one.

**Reply by blinkfrog** [reply]
Thank you for reply. Yes, yesterday, when I updated AdaptiveCpp, I noticed commits which renamed many things like this. However, when I use `-DACPP_DEBUG_LEVEL` option, I am getting this warning:
```
CMake Warning:
  Manually-specified variables were not used by the project:

    ACPP_DEBUG_LEVEL
```
And I don't get a similar warning when using `-DHIPSYCL_DEBUG_LEVEL`.

**Reply by blinkfrog** [reply]
Just in case, this is the version I compiled:
```
commit b5aa0c566817c2963525988cc356ef7b587e0cdf (HEAD -> develop, origin/develop, origin/HEAD)
Merge: 52fdcad0 35a512f1
Author: Aksel Alpay <aksel.alpay@uni-heidelberg.de>
Date:   Wed Aug 7 21:34:32 2024 +0200
```

**Reply by illuhad** [reply]
You're right, it seems that we didn't rename `HIPSYCL_DEBUG_LEVEL` after all. Sorry for that, I was mistaken. So `-DHIPSYCL_DEBUG_LEVEL` is still expected to work. Are you also compiling your application with `-DHIPSYCL_DEBUG_LEVEL=0`? Errors are sometimes also emitted in the headers; in this case it would not be sufficient to only compile AdaptiveCpp with `-DHIPSYCL_DEBUG_LEVEL=0` - you would also have to set this macro when building your application.

If the error is part of the AdaptiveCpp asynchronous error log (i.e. there is `============== hipSYCL error report ============== ` before it), then it is not subject to the `HIPSYCL_DEBUG_LEVEL`, but to the `async_handler` mechanism in SYCL.

**Reply by blinkfrog** [reply]
Thank you. Yes, I also have tried to add `-DHIPSYCL_DEBUG_LEVEL=0` when compiling application, but this still doesn't help.

This is the full output of my program, which attempts to allocate more memory than is available on device (to check how sysmem fallback option works in Nvidia drivers):

```
[AdaptiveCpp Warning] backend_loader: Could not load library: C:\adaptivecpp\bin\hipSYCL\rt-backend-hip.dll with: 126
Attempting to allocate 9 GB array on device...
[AdaptiveCpp Error] from D:/source/AdaptiveCpp/src/runtime/cuda/cuda_allocator.cpp:31 @ allocate(): cuda_allocator: cudaMalloc() failed (error code = CUDA:2)
Memory allocation failed!
Caught asynchronous SYCL exception: from D:/source/AdaptiveCpp/src/runtime/cuda/cuda_allocator.cpp:31 @ allocate(): cuda_allocator: cudaMalloc() failed (error code = CUDA:2)
```

Note the very first warning. It appears with any debug level, 0 or 1. While, when I compile this program with my previously built AdaptiveCpp (which was built with debug level 1), there is no this warning.

It looks like `HIPSYCL_DEBUG_LEVEL` has no any effect in the current version of AdaptiveCpp. I'll try to recompile previous version with DEBUG LEVEL = 0 to compare it with the recent version.

**Reply by blinkfrog** [reply]
Hm. This all is a bit different.
So, I tested older (May) version of AdaptiveCpp and the recent. Behavior of both is actually identical. `-DHIPSYCL_DEBUG_LEVEL=1` works correctly in both, setting debug level to errors only, and disabling warnings. However, `-DHIPSYCL_DEBUG_LEVEL=0` has no effect, both warnings and errors are displayed in both older and the recent AdaptiveCpp.

**Reply by blinkfrog** [reply]
If I use `-DHIPSYCL_DEBUG_LEVEL=0` to build AdaptiveCpp, then build.ninja file has "-DHIPSYCL_DEBUG_LEVEL=2" entries after configuring the build.

**Comment by nilsfriess** [comment]
I think there are actually multiple issues here. The first is that in the CMakeLists.txt we check
```cmake
if(NOT HIPSYCL_DEBUG_LEVEL)
```
and if that's true, we set a default value (2 for a release build). However, passing `-DHIPSYCL_DEBUG_LEVEL=0` will still make this `if` condition evaluate to true, so what we need to check is 
```cmake
if(NOT DEFINED HIPSYCL_DEBUG_LEVEL)
```
I think this is why you see "-DHIPSYCL_DEBUG_LEVEL=2" in build.ninja.

But even with this change, the value of `HIPSYCL_DEBUG_LEVEL` is actually never propagated to `include/hipSYCL/common/debug.hpp`, so if you do something like
```c++
#include <sycl/sycl.hpp>
#include <iostream>

int main() {
  std::cout << HIPSYCL_DEBUG_LEVEL << "\n";
}
```
then this will still output 2. We already have the file `include/hipSYCL/common/config.hpp.in`, I think we should just add something 
```c++
#define HIPSYCL_DEBUG_LEVEL @HIPSYCL_DEBUG_LEVEL@
```

But maybe this whole problem just indicates that we actually need multiple DEBUG_LEVEL variables? I guess in some cases you'd want verbose output when compiling but don't want verbose output in the actual program (or vice versa), although you could already achieve that by setting environment variables before compiling or running, respectively. Would probably be overkill to introduce two different debug level concepts, but we should fix the overall issue that HIPSYCL_DEBUG_LEVEL is not propagated from CMake -> SYCL headers.

What do you think @illuhad ?

**Comment by blinkfrog** [comment]
Closing as this was fixed in https://github.com/AdaptiveCpp/AdaptiveCpp/pull/1846, and also can be controlled with https://github.com/AdaptiveCpp/AdaptiveCpp/pull/1848
